File: rust/src/_49_group_anagrams.rs
/*
 * @lc app=leetcode id=49 lang=rust
 *
 * [49] Group Anagrams
 */

// @lc code=start
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map: HashMap<[usize; 26], Vec<String>> = HashMap::new();
        for s in strs {
            let mut freq = [0; 26];
            s.chars().for_each(|c| {
                let i = (c as u8 - b'a') as usize;
                freq[i] += 1;
            });
            map.entry(freq).or_default().push(s);
        }
        map.values().cloned().collect()
    }
}
// @lc code=end

struct Solution;

File: golang/src/_49_group_anagrams.go
/*
 * @lc app=leetcode id=49 lang=golang
 *
 * [49] Group Anagrams
 */

// @lc code=start
func groupAnagrams(strs []string) [][]string {
	groups := make(map[[26]byte][]string)
	for _, s := range strs {
		var key [26]byte
		for _, c := range s {
			key[c-'a']++
		}
		groups[key] = append(groups[key], s)
	}
	result := make([][]string, 0, len(groups))
	for _, group := range groups {
		result = append(result, group)
	}
	return result
}

// @lc code=end
